plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs

```diff
@@ -158,8 +156,50 @@ class ExMemoryUsage a where
     -- Inlining the implementations of this method gave us a 1-2% speedup.
     memoryUsage :: a -> CostRose
 
-instance (ExMemoryUsage a, ExMemoryUsage b) => ExMemoryUsage (a, b) where
-    memoryUsage (a, b) = CostRose 1 [memoryUsage a, memoryUsage b]
+{- Note [Alternative memory usage instances].  The `memoryUsage` function provides
+ a measure of the size of an object for costing purposes, the idea being that
+ the time taken to execute a builtin, and the memory used to contain its result,
+ will depend on the size of the inputs.  The name `memoryUsage` is perhaps a
+ misnomer: it was originally supposed to measure (in 64-bit words) the heap
+ space required to store an object, but this is not always the correct measure
+ to use.  For example, the time taken by `AddInteger` or `MultiplyInteger` will
+ depend on the logarithms of the inputs (and the logarithm is proportional to
+ the memory occupied by the inputs), and the memory occupied by the result will
+ be some function of the memory occupied by the inputs, so for these functions
+ the actual memory usage is a sensible size measure.  However, calling
+ `replicateByte n b` function allocates a number of bytes which is equal to the
+ actual value of `n`, which will be exponentially greater than the memoory
+ occupied by `n`, so this case the memory usage is not a sensible size measure.
+ In most cases the default `memoryUsage` function returns the actual memory
+ usage, butto deal with cases like `replicateByte` we occasionally use newtype
+ wrappers which define a different size measure (see `IntegerCostedLiterally`
+ below).  Polymorhpic types require some care though: see Note [ExMemoryUsage
+ for polymorphic types].
+-}
+
+{- Note [ExMemoryUsage for polymorphic types].  For polymorphic types such as
+ `pair, `list`, and `array` DO NOT use newtype wrappers to define alternative
+ size measures.  The denotations of functions which take polymorphic arguments
+ use `SomeConstant` and this willl ignore newtype wrappers and will only use the
+ default `memoryUsage` function, which could lead to unexpected results.
+ Furthermore, actual memory usage is typically not a good size measure for
+ polymorphic arguments: the time taken to process a list, for example, will
+ typically depend only on the length of the list and not the size of the
+ contents.  Currently all such functions are parametrically polymorphic and only
+ manipulate pointers without inspecting the contents of their polymorphic
+ arguments, so it is reasonable to use size measures which depend only on the
+ surface structure of polymorphic objects. -}
+
+instance ExMemoryUsage (a, b) where
+    memoryUsage _ = singletonRose 1
+    {-# INLINE memoryUsage #-}
```

Comment:
The indentation is inconsistent across these instances: some use 4 spaces, some 3.
I suggest raking the whole file with `fourmolu`

```suggestion
  memoryUsage _ = singletonRose 1
  {-# INLINE memoryUsage #-}
```

Reply:
>  I suggest raking the whole file with fourmolu
I'll do that before merging.  It changes almost every line though, so I won't do it right now.

Reply:
Thanks!